engine/src/render/cascaded_shadow_maps_rasterizer/plugin.cpp

```diff
@@ -0,0 +1,314 @@
+
+                    // Send the PerScene data to the GPU.
+                    PerScene perScene;
+
+                    float cameraZFar = 1000.0F;
+                    if (perspectiveCamera.contains())
+                    {
+                        cameraZFar = perspectiveCamera.value().zFar;
+                    }
+                    else if (orthographicCamera.contains())
+                    {
+                        cameraZFar = orthographicCamera.value().zFar;
+                    }
+
+                    float maxFar = caster.maxDistance == 0 ? cameraZFar : caster.maxDistance;
+                    for (int i = 0; i < caster.numSplits; i++)
+                    {
+                        float near = std::max(caster.nearDistance,
+                                              maxFar * caster.splitDistances.at(static_cast<unsigned long>(i)));
+                        float far = maxFar * (i == caster.numSplits - 1
+                                                  ? 1.0F
+                                                  : caster.splitDistances.at(static_cast<unsigned long>(i) + 1));
+
+                        auto cameraView = glm::inverse(cameraLocalToWorld.mat);
+
+                        glm::mat4 cameraProj;
+                        if (perspectiveCamera.contains())
+                        {
+                            cameraProj = glm::perspective(glm::radians(perspectiveCamera.value().fovY),
+                                                          (float(gBuffer.size.x) * drawsTo.viewportSize.x) /
+                                                              (float(gBuffer.size.y) * drawsTo.viewportSize.y),
+                                                          near, far);
+                        }
+                        else if (orthographicCamera.contains())
+                        {
+                            const auto& ortho = orthographicCamera.value();
+                            float aspect = (static_cast<float>(gBuffer.size.x) * drawsTo.viewportSize.x) /
+                                           (static_cast<float>(gBuffer.size.y) * drawsTo.viewportSize.y);
+                            if (ortho.axis == OrthographicCamera::Axis::Vertical)
+                            {
+                                cameraProj = glm::ortho(-ortho.size * aspect, ortho.size * aspect, -ortho.size,
+                                                        ortho.size, near, far);
+                            }
+                            else
+                            {
+                                cameraProj = glm::ortho(-ortho.size, ortho.size, -ortho.size / aspect,
+                                                        ortho.size / aspect, near, far);
+                            }
+                        }
+
+                        std::vector<glm::vec4> cameraFrustumCorners;
+                        getCameraFrustumCorners(cameraProj * cameraView, cameraFrustumCorners);
```

Comment:
```suggestion
                        getCameraFrustumCorners(camera.projection * cameraView, cameraFrustumCorners);
```
You don't need to calculate the projection yourself right? Why not use the field in the Camera component?
Then this whole code wouldn't need references to specific camera types, right? I might be missing something in the math

Reply:
The problem is that the way the camera frustum is split for each cascade is by changing the near/far planes in the projection.

On a second thought, perhaps `getCameraFrustumCorners` can be modified to convert different Z coordinates rather than just the corners. This way the frustum could be split without recalculating the projection matrix.